File: 2015/AI/AIClient_JAVA/src/world/Character.java
/* ------------------------------------------------------------------------------
 ** _________   _________      ________    _____      _____  ___________ _________
 ** \_   ___ \ /   _____/     /  _____/   /  _  \    /     \ \_   _____//   _____/
 ** /    \  \/ \_____  \     /   \  ___  /  /_\  \  /  \ /  \ |    __)_ \_____  \ 
 ** \     \____/        \    \    \_\  \/    |    \/    Y    \|        \/        \
 **  \______  /_______  /     \______  /\____|__  /\____|__  /_______  /_______  /
 **         \/        \/             \/         \/         \/        \/        \/ 
 **
 ** Character.java
 ** A character
 **
 ** Author: Samuel-Ricardo Carriere
 ** ------------------------------------------------------------------------------*/

package world;

import mathUtils.Direction;
import mathUtils.Vector2;
import event.DropMineEvent;
import event.MoveCharacterEvent;
import event.QueueController;
import event.ShootMissileEvent;

/**
 * Class that contain all the information about the character
 */
public class Character {
	private Vector2 position;
	private int id;
	private int life;
	private Mine mine = new Mine();
	private Missile missile = new Missile();

	public Character(int id) {
		this.id = id;
		this.position = new Vector2(0, 0);
		this.life = 3;
	}

	public void updateInfo(Vector2 position) {
		this.position = position;
	}
	
	public void hitByMine() {
		if (isAlive()) {
			--life;
		}
	}

	public void hitByMissile() {
		if (isAlive()) {
			--life;
		}
	}

	public void mineHit() {
		mine.hit();
	}

	public void missileHit() {
		missile.hit();
	}
	
	/**
	 * Send the order to move the character to a certain position
	 * 
	 * Exemple :
	 *		character.move(new Point(5,5));
	 */	
	public void move(Vector2 position) {
		QueueController queueController = QueueController.getInstance();
		MoveCharacterEvent event = new MoveCharacterEvent(id, position);
		queueController.addOutgoingEvent(event);
	}

	/**
	 * Send the order to drop a mine at the character position
	 * if the mine is ready
	 * 
	 * Exemple :
	 *		character.dropMine();
	 */	
	public void dropMine() {
		if (mine.isReady()) {
			QueueController queueController = QueueController.getInstance();
			DropMineEvent event = new DropMineEvent(id);
			queueController.addOutgoingEvent(event);
			mine.drop();
		}
	}

	/**
	 * Send the order to shoot a missile on a certain direction
	 * if the missile is ready
	 * 
	 * Exemple :
	 *		character.shootMissile(Missile.Direction.UP);
	 */	
	public void shootMissile(Direction direction) {
		if (missile.isReady()) {
			QueueController queueController = QueueController.getInstance();
			ShootMissileEvent event = new ShootMissileEvent(id,
					direction.ordinal());
			queueController.addOutgoingEvent(event);
			missile.shoot();
		}
	}

	/**
	 * Return the id of this team
	 * 
	 * @return the id of this team
	 */
	public Vector2 getPosition() {
		return position;
	}

	/**
	 * Check if the character still have some life point
	 * 
	 * @return true if the character is alive, else false
	 */
	public boolean isAlive() {
		return life > 0;
	}

	/**
	 * Return the missile of the character
	 * 
	 * @return the missile
	 */
	public Missile getMissile() {
		return missile;
	}
}

File: 2015/AI/AIClient_Python/src/world/Character.py
'''
Created on Dec 20, 2014

@author: scarriere
'''

from event.DropMineEvent import DropMineEvent
from aiclient.Singleton import Singleton
from event.QueueController import QueueController
from world.Missile import Missile
from mathUtils.Vector2 import Vector2
from event.MoveCharacterEvent import MoveCharacterEvent
from event.ThrowMissileEvent import ThrowMissileEvent, Direction


class Character(object):
    '''
    Class that contain all the informations about the Character
        (You can't use any of the functions or variables that start with an "_")
    '''
    _queueController = Singleton(QueueController)
    
    def __init__(self, characterId, position):
        self._characterId = characterId
        self.position = Vector2(position.x, position.y)
        '''
        The current position of the character
            (see :class:`.Vector2`)
        '''
        self.mineReady = True
        '''
        Is the mine ready to be use
            (see :class:`.bool`)
        '''
        self.missile = Missile(Vector2(0,0))
        '''
        The missile of the character
            (see :class:`.Missile`)
        '''
        self.life = 3
        '''
        The life total of the character
            (see :class:`.int`)
        '''
    
    def _updatePosition(self, position):
        self.position.x = position.x
        self.position.y = position.y
        
    def _hitByMine(self):
        if self.isAlive():
            self.life -= 1
    
    def _mineHit(self):
        self.mineReady = True
    
    def _hitByMissile(self):
        if self.isAlive():
            self.life -= 1
        
    def _missileHit(self):
        self.missile.isReady = True

    def goTo(self, position):
        '''
        Send the order to move the character
        to a certain position(:class:`.Vector2`)
            
        Exemple::
        
            aCharacter.goTo(Vector2(6,7))
        '''
        event = MoveCharacterEvent(self._characterId, position.x, position.y)
        self._queueController.outEvents.put(event)
        
    def dropMine(self):
        '''
        Send the order to drop a mine on the current position(:class:`.Vector2`)
            if the mine is ready
            
        Exemple::
        
            aCharacter.dropMine()
        '''
        if(self.mineReady):
            event = DropMineEvent(self._characterId)
            self._queueController.addOutgoingEvent(event)
            self.mineReady = False

    def shootMissile(self, direction):
        '''
        Send the order to shoot a missile on a certain direction(:class:`.Direction`)
            if the missile is ready
            
        Exemple::
        
            aCharacter.shootMissile(Direction.UP)
        '''
        if(self.missile.isReady):
            event = ThrowMissileEvent(self._characterId, direction)
            self._queueController.addOutgoingEvent(event)
            self.missile.isReady = False
        
    def isAlive(self):
        '''
        Check if the character still have some life point
            
        Exemple::
        
            checkIfAlive = aCharacter.isAlive()
        '''
        return self.life > 0
    
    def getPosition(self):
        '''
        Return the position of the character
            
        Exemple::

            position = aCharacter.getPosition()
        '''
        return self.position
